content/scala-cli.md

```diff
@@ -0,0 +1,150 @@
+
+We have described the motivation, syntax and implementation basis in the [dedicated pre-SIP](https://contributors.scala-lang.org/t/pre-sip-using-directives/5700). Currently, we recommend to write using directives as comments, so making them part of the language specification is not necessary at this stage. Moreover, the new `scala` command could ignore using directives in the initial version, however we strongly suggest to include comment-based using directives from the start. 
+
+### Compatibility
+
+Adopting Scala CLI as the new `scala` command, as is, will change some of the behaviour of today's scripts. Some examples:
+
+- `scala repl` needs to be run to start a REPL instead of just `scala`
+- with `scala compile` and `scala doc`, some of the more obscure (or brand new) compile options need to be prefixed with `-O`
+- Scala CLI recognizes tests based on the extension used (`*.test.scala`) so running `scala compile a.scala a.test.scala` will only compile `a.scala`
+- Scala CLI has its own versioning scheme, that is not related to the Scala compiler. Default version used may dynamically change when new Scala version is released.
+- By default, Scala CLI manages its own dependencies (e.g. scalac, zinc, Bloop) and resolves them lazily. This means that the first run of Scala CLI resolves quite some dependencies. Moreover, Scala CLI periodically checks for updates, new defaults accessing online resources (but it is not required to work, so Scala CLI can work in offline environment once setup)
+- Scala CLI can also be configured via using directives. Command line options have precedence over using directives, however using directives override defaults. Compiling a file starting with `//> using scala 2.13.8`, without providing a Scala version on the command line, will result in using `2.13.8` rather than the default Scala version. We consider this a feature. However, technically, this is a breaking change.
+
+### Other concerns
+
+Scala CLI brings [using directives](https://scala-cli.virtuslab.org/docs/guides/using-directives) and  [conventions to mark the test files](https://scala-cli.virtuslab.org/docs/commands/test#test-sources). We are not sure if both can be accepted as a part of this SIP or we should have seperate SIPs for both (we have opened a [pre-SIP for using directives](https://contributors.scala-lang.org/t/pre-sip-using-directives/5700/15))
+
+Scala CLI is an ambitious project and may seem hard to maintain in the long-run. // TODO
+
+
+### Open questions
+
+The main open question for this proposal is wich commands/features should be included by default in the `scala` command. Another aspect is the release cadence: should the new `scala` command follow the current release cadence for Scala CLI (every 2 weeks) or stick to Scala one (every 6 weeks)?
```

Comment:
```suggestion
The main open question for this proposal is which commands/features should be included by default in the `scala` command. Another aspect is the release cadence: should the new `scala` command follow the current release cadence for Scala CLI (every 2 weeks) or stick to Scala one (every 6 weeks)?
```

Reply:
Given that Scala CLI will dynamically resolve various Scala versions, I don't see the point in tying its release schedule to that of dotc.

Reply:
I am also all for the independent release scheme but @julienrf and others shared some concerns about this, and that is why I have included it as an open question.

Reply:
Currently, the setup instructions tell you to run `scala -version` to test your setup:

![Screenshot from 2022-07-20 13-38-30](https://user-images.githubusercontent.com/332812/179972935-e2a20c0e-8a4d-4b48-8bb0-a63a41a767e5.png)

What do we want `scala -version` to return when `scala` is implemented via `scala-cli`? I thought it could make sense to build a `scala` command with every release of Scala, and which would default to the Scala version of that release.

Otherwise, we can decide that the version of `scala` has nothing to do with an actual version of Scala, and in that case, we should document it explicitly.

Reply:
IMO, `scala -version` should show the version of Scala that will be used if I use `scala` (i.e., `scala repl`) or `scala compile` in the current directory (which means taking `using` clauses into account). The version of the launcher is irrelevant in 99% of use cases, and probably 100% of *beginner* use cases.

For the Python lovers, compare with `py -version`/`py -3 -version` which show "Python 3.10.5" and `py -2 -version` which shows "Python 2.7.18": I get the version of Python that will be used if I run `py`/`py -3` or `py -2`. The version of `py` itself is irrelevant.

Reply:
Good point, I created an issue for it: https://github.com/VirtusLab/scala-cli/issues/1250

Reply:
> After a discussion with Sébastien, here is a suggestion about this point: when someone invokes `scala -version`, simply prints both the version of the runner (ie, scala-cli), and the _current_ version of Scala. The _current_ version of Scala would depend on the working directory (e.g., if there is a Scala project that configures a specific Scala version with a “using” directive, that version should be printed).
> 
> Example:
> 
> ```
> $ scala -version .
> Scala code runner version: 1.2.3
> Scala version (set in ./my-program.scala:2): 3.1.3
> ```
> 
> Or:
> 
> ```
> $ scala -version
> Scala code runner version: 1.2.3
> Default Scala version: 3.2.0
> ```

What is the motivation for mixing up the concerns of the version of `scala-cli` and "inspecting" a project/file? I suspect having multiple versions displayed in the output would create some confusion for new users. 

It would be useful for the default scala version (`--scala-version` option) to  be displayed in the output of `--help` for the relevant subcommands (e.g., `run`, `compile`, etc...). Currently, there's no default shown.

```bash
$> scala-cli compile --help | grep scala-version
  -S, --scala, --scala-version version           Set the Scala version
```

However, a useful error message is displayed if an unsupported value is provided.

```bash
$> scala-cli run --scala-version 7 Example.scala
[error]  Cannot find matching Scala version for '7'
You can only choose one of the 3.x, 2.13.x, and 2.12.x. versions.
The latest supported stable versions are 3.1.3, 2.13.8, 2.12.16.
In addition, you can request compilation with the last nightly versions of Scala,
by passing the 2.nightly, 2.12.nightly, 2.13.nightly, or 3.nightly arguments.
Specific Scala 2 or Scala 3 nightly versions are also accepted.
```

Ideally, the defaults of other command line flags, such as `--jvm`, would be communicated in the output of `--help` as well.

I agree with your suggestion that "inspecting" a project would be a welcomed addition. Perhaps this use case would be better as a new subcommand that could provide more information about a project/file. For example, the configured scala version, the `--scalac-options` used, the resolved dependencies, or a list of main classes. Having an "inspect" functionality  decoupled from `--version` and captured in a subcommand would also enable adding more fine grained information about a project/file over time. Overloading `--version` with this info would probably not age well. 

Currently, I find `scala-cli` sometimes a little tricky to debug when I've configured it incorrectly (e.g., main class), or for cases where it's unclear what default values are being used by `scala-cli`. It could be useful to get feedback from the community to evaluate if an "inspect" functionality could improve the overall user experience.

Overall, I'm finding `scala-cli` to be a very positive addition to the scala ecosystem.

Reply:
> What is the motivation for mixing up the concerns of the version of scala-cli and "inspecting" a project/file?

I guess someone who installed Scala would be very likely to try `scala -version` to know which version of Scala is installed. But since `scala` will be a runner able to run any version of Scala, it makes sense to also show to the user the version of the CLI that is installed.

A similar situation is when someone starts a REPL session. We show both the JDK version and the Scala version because both are useful.